### PR TITLE
6550 and 18776 -  Need to see Last Credit Company Selected For Last Credit Company bill (in Patient Details). When Patient came for New visit.

### PR DESCRIPTION
**Pull request by damithdeshan98**
## Description

  When a patient selects **Credit** as the payment method in OPD Billing, there is no way to quickly re-use a credit company that was previously used for the same patient. The cashier has to manually search
  and re-enter the company, policy number, and reference number every time.

  This feature queries the patient's past credit payments and displays them in a table so the cashier can select a previous company with one click to auto-fill the credit payment fields.

  ---

  ## Files Changed

  | File | Change |
  |---|---|
  | `OpdBillController.java` | Added JPQL query to fetch previous credit companies + `selectCreditCompany()` method |
  | `CreditCompanyDetailsDto.java` | New DTO — holds `companyId`, `companyname`, `policyNo`, `referenceNo` |
  | `credit_with_last_company.xhtml` | New composite component — shows previous company table with Select button |
  | `opd_bill_ac.xhtml` | Integrated the new composite component inside Credit payment panel |

  ---

  ## Configuration Options

  | Key | Default Value | Description |
  |---|---|---|
  | `Display the past Credit Company List` | `false` | Enables/disables the previous credit company table in OPD Billing. When `false`, the standard credit input is shown. |
  | `How many years should you search back to find a credit company?` | `5` | Number of past years to search for previous credit payments for the patient. |

  ---

  ## UI Changes

  - When **Credit** payment method is selected and the config option is enabled, a **"Previous Credit Company"** panel appears above the credit input fields.
  - The panel is **hidden when the list is empty** (no previous credit history for the patient).
  - The table displays: `#`, `Company Name`, `Policy No`, `Reference No`, and a `Select` button.
  - Clicking **Select** auto-fills the Credit Company, Policy No, and Reference No fields without triggering form validation.

  ---

  ## How to Test

  1. **Enable the config option**
     - Go to Application Configuration
     - Set `Display the past Credit Company List` = `true`

  2. **Create test data**
     - Create an OPD bill for a patient using **Credit** payment (e.g. AIA Insurance, Policy: `POL-001`, Ref: `REF-001`)
     - Complete and save the bill

  3. **Test the feature**
     - Start a new OPD bill for the **same patient**
     - Select **Credit** as the payment method
     - Verify the **"Previous Credit Company"** panel appears with the previous company listed
     - Click **Select** on a row
     - Verify the Credit Company, Policy No, and Reference No fields are auto-filled correctly

  4. **Test edge cases**
     - Patient with **no previous credit history** → panel should not appear
     - Patient with **multiple different credit companies** → all should appear (grouped by company)
     - Change `How many years should you search back to find a credit company?` to `1` → only records within the last year should appear

## Summary by CodeRabbit

* **New Features**
  * OPD Credit payments can show and let you pick from previously used credit companies (configurable).
  * Credit entry UI now shows historical company rows, plus policy/reference and memo fields, and an admin link to add companies.
  * Added a small data structure to support credit-company details.

* **Bug Fixes**
  * Removed a debug log that could appear during billing.

* **Chores**
  * Added a utility to compute past dates used by billing flows.

### CODE REVIEW

File: src/main/java/com/divudi/bean/opd/OpdBillController.java [hunk 1]
```diff
@@ -4719,10 +4723,64 @@ public void listnerForPaymentMethodChange() {
...
+                            + " p.referenceNo "
+                            + " ) "
+                            + "FROM Payment p "
+                            + "WHERE p.retired = :ret "
+                            + "AND p.paymentMethod =:method "
+                            + "AND p.bill.cancelled =:can "
+                            + "AND p.bill.patient =:pt "
+                            + "AND p.creditCompany IS NOT NULL "
+                            + "AND p.createdAt between :fDate and :tDate "
+                            + "GROUP BY p.creditCompany, p.creditCompany.id, p.creditCompany.name, p.policyNo, p.referenceNo";
```
[the remark this answers is not in the document]
Comment (damithdeshan98) [reply src/main/java/com/divudi/bean/opd/OpdBillController.java:4745]: No Need

File: src/main/java/com/divudi/core/util/CommonFunctions.java [hunk 2]
```diff
@@ -467,6 +467,12 @@ public static Date getStartOfDay(Date date) {
     public static Date getEndOfDay() {
         return getEndOfDay(new Date());
     }
+    
+    public static Date getPreviousDate(int yearsBack) {
+        Calendar calendar = Calendar.getInstance();
+        calendar.add(Calendar.YEAR, -yearsBack);
+        return calendar.getTime();
```
[the remark this answers is not in the document]
Comment (damithdeshan98) [reply src/main/java/com/divudi/core/util/CommonFunctions.java]: Done

File: src/main/webapp/resources/paymentMethod/credit_with_last_company.xhtml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                        <p:commandButton
+                            value="Select"
+                            icon="pi pi-check"
+                            process="@this"
+                            update="@form"
+                            actionListener="#{cc.attrs.controller.selectCreditCompany(cmp)}">
+                        </p:commandButton>
+                    </p:column>
+                </p:dataTable>
+            </p:panel>
```
[the remark this answers is not in the document]
Comment (damithdeshan98) [reply src/main/webapp/resources/paymentMethod/credit_with_last_company.xhtml]: Fixed

File: src/main/java/com/divudi/core/data/dto/CreditCompanyDetailsDto.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public void setCompanyname(String companyname) {
+        this.companyname = companyname;
+    }
+
+    @Override
+    public String toString() {
+        return "InstitutionDto{id=" + companyId + ", name='" + companyname + "'}";
+    }
```
[the remark this answers is not in the document]
Comment (damithdeshan98) [reply src/main/java/com/divudi/core/data/dto/CreditCompanyDetailsDto.java:63]: Done
